cdc/sink/mq.go

```diff
@@ -164,19 +173,56 @@ func (k *mqSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowCha
 	return nil
 }
 
+// FlushRowChangedEvents is thread-safety
 func (k *mqSink) FlushRowChangedEvents(ctx context.Context, tableID model.TableID, resolvedTs uint64) (uint64, error) {
-	if checkpointTs, ok := k.tableCheckpointTs[tableID]; ok && resolvedTs <= checkpointTs {
+	var checkpointTs uint64
+	v, ok := k.tableCheckpointTsMap.Load(tableID)
+	if ok {
+		checkpointTs = v.(uint64)
+	}
+	if resolvedTs <= checkpointTs {
 		return checkpointTs, nil
 	}
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case k.resolvedBuffer <- resolvedTsEvent{
+		tableID:    tableID,
+		resolvedTs: resolvedTs,
+	}:
+	}
+	k.statistics.PrintStatus(ctx)
+	return checkpointTs, nil
+}
 
+// bgFlushTs flush resolvedTs to workers and flush the mqProducer
+func (k *mqSink) bgFlushTs(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg := <-k.resolvedBuffer:
+			resolvedTs := msg.resolvedTs
+			err := k.flushTsToWorker(ctx, resolvedTs)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = k.mqProducer.Flush(ctx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			k.tableCheckpointTsMap.Store(msg.tableID, resolvedTs)
```

Comment:
What if a table is removed between line 213 and 214?

Reply:
Since CDC does not guarantee exactly onece semantic, it won't cause any poblem here.

Let say `capture1` has got a table name `t1` and `t1.resolvedTs = 10`, `t1.checkpointTs = 5`. If `t1` was moved to `capture2` here, then `t1` will be replicated from `t1.checkPointTs = 5` in `capture2` so that the `data [5, 10]` would be sent twice, but no data will be lost.